### PR TITLE
Jetpack Sync: Don't sync users with user_level 0 on "initial_sync"

### PR DESCRIPTION
**Pull request by lezama**
We sync users on plugin upgrade so we have the information to populate subscription and notification posts. We don't need users with user_level 0 at this point. 

cc @gravityrail @enejb

### CODE REVIEW

**Comment by kraftbj** [comment]
Do we want to use `user_levels`? They've been deprecated in WP since 3.0.

**Comment by gravityrail** [comment]
I think for safety we should check that the number of users at that level is not more than, say, 500.

We could either include `ORDER BY ID ASC LIMIT 500`, or perhaps we could count the users before retrieving the IDs, and if it's > 500 we can just not sync users at all? Or only sync some smaller group like admins.

**Comment by lezama** [comment]
> Do we want to use user_levels? They've been deprecated in WP since 3.0.

I guess I need to find a better way to get those ids

**Comment by gravityrail** [comment]
It seems like what we really need to do here is find the list of roles that have the capability to author and publish posts, and then select users with those roles.

File: sync/class.jetpack-sync-actions.php [hunk 1]
```diff
@@ -135,11 +135,16 @@ static function send_data( $data, $codec_name, $sent_timestamp, $queue_id ) {
 		return $rpc->getResponse();
 	}
 
+	static function get_user_ids_for_initial_sync() {
+		global $wpdb;
+		return $wpdb->get_col( "SELECT user_id FROM $wpdb->usermeta WHERE meta_key = '{$wpdb->base_prefix}user_level' AND meta_value > 0" );
```
Comment (enejb) [inline sync/class.jetpack-sync-actions.php]: I think we need to do set a limit here and chunk process it in batches. Like we currently do for users full sync.

**Comment by lezama** [comment]
Thanks for the comments! pushed a change to use `WP_User_Query`

This is what it builds:

```SELECT SQL_CALC_FOUND_ROWS wp_users.ID FROM wp_users INNER JOIN wp_usermeta ON ( wp_users.ID = wp_usermeta.user_id ) WHERE 1=1 AND ( 
  ( wp_usermeta.meta_key = 'wp_user_level' AND wp_usermeta.meta_value != '0' )```

File: tests/php/sync/test_class.jetpack-sync-full.php [hunk 2]
```diff
@@ -984,6 +984,42 @@ function test_sync_call_ables_does_not_modify_globals() {
...
+					'network_options' => true,
+					'functions' => true,
+					'constants' => true,
+					'users' => Jetpack_Sync_Actions::get_initial_sync_user_config(),
+				)
+			)
+		);
+	}
+
+	function _do_cron() {
```
Comment (gravityrail) [inline tests/php/sync/test_class.jetpack-sync-full.php]: You can delete this, it's no longer needed

File: sync/class.jetpack-sync-actions.php [hunk 3]
```diff
@@ -135,11 +136,19 @@ static function send_data( $data, $codec_name, $sent_timestamp, $queue_id ) {
 		return $rpc->getResponse();
 	}
 
+	static function get_initial_sync_user_config() {
+		$user_query = new WP_User_Query( array( 'who' => 'authors', 'fields' => 'ID' ) );
```
Comment (enejb) [inline sync/class.jetpack-sync-actions.php]: Could we set a limit to WP_User_Query to be MAX_INITIAL_SYNC_USER + 1. 
This way the query wouldn't have to do as much work.

Comment (enejb) [inline sync/class.jetpack-sync-actions.php]: Also I think we would want to sync contributors as well.

**Comment by gravityrail** [comment]
Merged to 4.2

**Pull request merged**